script/tool/lib/src/common/cmake.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file/file.dart';
+import 'package:flutter_plugin_tools/src/common/core.dart';
+import 'package:platform/platform.dart';
+
+import 'process_runner.dart';
+
+const String _cmakeCacheCommandKey = 'CMAKE_COMMAND:INTERNAL';
+
+/// A utility class for interacting with CMake projects.
+class CMakeProject {
+  /// Creates an instance that runs commands for [project] with the given
+  /// [processRunner].
+  CMakeProject(
+    this.flutterProject, {
+    required this.buildMode,
+    this.processRunner = const ProcessRunner(),
+    this.platform = const LocalPlatform(),
+  }) {
+    // On Linux 'cmake' is expected to be in the path, so doesn't need to
+    // be lookup up and cached (see [getCMakeCommand]).
+    if (platform.isLinux) {
+      _cmakeCommand = 'cmake';
+    }
+  }
+
+  /// The directory of a Flutter project to run Gradle commands in.
+  final Directory flutterProject;
+
+  /// The [ProcessRunner] used to run commands. Overridable for testing.
+  final ProcessRunner processRunner;
+
+  /// The platform that commands are being run on.
+  final Platform platform;
+
+  /// The build mode (e.g., Debug, Release).
+  ///
+  /// This is a constructor paramater because on Linux many properties depend
+  /// on the build mode since it uses a single-configuration generator.
+  final String buildMode;
+
+  String? _cmakeCommand;
```

Comment:
nit: I don't believe you need an instance variable for this.  The code would be easier to maintain if it was just defined in `getCMakeCommand`

Reply:
I'm not sure I follow; the purpose of a cache isn't simplicity, it's efficiency. Are you saying it's bad to cache the results of parsing a file?

Reply:
Yep, mutable data and caching can lead to bugs and should be avoided.  If you want to cache for performance I'd recommend a memoization pattern:

```dart
lazy int _memoizedGetHash => _calculateHash();
int getHash() {
  return _memoizedGetHash;
}
```

Reply:
I'm not familiar with `lazy`, and DartPad doesn't think it's valid Dart. Do you have more info on this?

Reply:
Sorry, got my syntax wrong, lazy values are 'late' in Dart:
```dart
late final int _memoizedGetHash => _calculateHash();
int getHash() {
  return _memoizedGetHash;
}
```

Reply:
Oh, nice! I had no idea `late` had that second meaning.